modules/core/src/smithy4s/http/internals/UrlFormDataDecoderSchemaVisitor.scala

```diff
@@ -0,0 +1,225 @@
+ */
+
+package smithy4s
+package http
+package internals
+
+import alloy.UrlFormFlattened
+import alloy.UrlFormName
+import smithy4s.codecs.PayloadPath
+import smithy4s.internals.SchemaDescription
+import smithy4s.schema._
+
+private[http] class UrlFormDataDecoderSchemaVisitor(
+    val cache: CompilationCache[UrlFormDataDecoder],
+    // These are used by AwsEc2QueryCodecs to conform to the requirements of
+    // https://smithy.io/2.0/aws/protocols/aws-ec2-query-protocol.html?highlight=ec2%20query%20protocol#query-key-resolution.
+    ignoreUrlFormFlattened: Boolean,
+    capitalizeStructAndUnionMemberNames: Boolean
+) extends SchemaVisitor.Cached[UrlFormDataDecoder]
+    with smithy4s.ScalaCompat {
+  compile =>
+
+  override def primitive[P](
+      shapeId: ShapeId,
+      hints: Hints,
+      tag: Primitive[P]
+  ): UrlFormDataDecoder[P] = {
+    val desc = SchemaDescription.primitive(shapeId, hints, tag)
+    Primitive.stringParser(tag, hints) match {
+      case Some(parser) =>
+        UrlFormDataDecoder.fromStringParser(desc)(parser)
+
+      case None =>
+        UrlFormDataDecoder.alwaysFailing(s"Cannot decode $desc from URL form")
+    }
+  }
+
+  override def collection[C[_], A](
+      shapeId: ShapeId,
+      hints: Hints,
+      tag: CollectionTag[C],
+      member: Schema[A]
+  ): UrlFormDataDecoder[C[A]] = {
+    val memberDecoder = compile(member)
+    val maybeKey =
+      if (ignoreUrlFormFlattened || hints.has[UrlFormFlattened]) None
+      else Option(getKey(member.hints, "member"))
+    cursor =>
+      maybeKey.fold(cursor)(cursor.down(_)) match {
+        case UrlFormCursor(_, Nil) =>
+          Right(tag.empty)
+
+        case UrlFormCursor(history, values) =>
+          // Collection members aren't necessarily primitives, and if they
+          // aren't, then there will be multiple values for the same
+          // index. One example is maps, which are encoded as collections of
+          // structs, e.g.
+          // foos.entry.1.key=a&foos.entry.1.value=1&foos.entry.2.key=b&foos.entry.2.value=2.
+          // That's why we have to group by index.
+          //
+          // We can't assume they were encoded in order. That's why we have to
+          // then sort by index.
+          val groupedAndSortedCursors = values
+            .collect {
+              case formData @ UrlForm.FormData(
+                    PayloadPath(PayloadPath.Segment.Index(index) :: _),
+                    _
+                  ) =>
+                index -> formData
+            }
+            .groupBy { case (index, _) =>
+              index
+            }
+            .toVector
+            .sortBy { case (index, _) =>
+              index
+            }
+            .map { case (index, indicesAndValues) =>
+              UrlFormCursor(
+                history,
+                indicesAndValues.map { case (_, value) => value }
+              )
+                .down(PayloadPath.Segment.Index(index))
+            }
```

Comment:
you could use `groupMap` to make it more efficient. It's available in 2.12 thanks to [scala-collection-compat](https://github.com/scala/scala-collection-compat/blob/a75be4fb207d42b8b3158ee4012ce05781cd51f8/compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala#L572-L583)

Reply:
👍 Done in https://github.com/disneystreaming/smithy4s/pull/1113/commits/e28d74db02953221ac20cd7307011f3f153f1b1d.

Reply:
And for 2.12 in https://github.com/disneystreaming/smithy4s/pull/1113/commits/b0ad4eb2eae62642fbfc6ce7d83e153e6774c6b1.